Support VPC security group for PowerVS clusters

dharaneeshvrd opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
Add support for VPC security group for PowerVS clusters

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
Fixes #1706 

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support VPC security group for PowerVS clusters
```

Prajyot-Parab commented:
/cc @Karthik-K-N

Karthik-K-N commented:
/cc @cjschaef

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+					return false, fmt.Errorf("error validating security group rule: %v", err)
+				}
+				if !match {
+					return false, nil
```
How about logging with V(3) with which security group rule did not match.

dharaneeshvrd replied:
It's really difficult to get that since it's not that straight forward. Let's keep it like this for now!

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+// validateSecurityGroupRule compares a specific security group's rule with the spec and existing security group's rule.
+func (s *PowerVSClusterScope) validateSecurityGroupRule(originalSecurityGroup vpcv1.SecurityGroup, direction infrav1beta2.SecurityGroupRuleDirection, protocol string, portMin, portMax int64, icmpCode, icmpType *string, remote infrav1beta2.SecurityGroupRuleRemote) (bool, error) {
+	var match bool
+	var err error
```
How about declaring them in return parameters?

dharaneeshvrd replied:
Done

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+				}
+			}
+		}
+		if match {
```
Is this should be !match? because we should validate all the rules, If the first rule matches we should not return right?

cjschaef replied:
Perhaps we do not return and attempt to validate only one rule matches.
I expect this function is attempting to verify only one SecurityGroupRule definition from `Spec` exists in the SecurityGroup. And the function (`validateSecurityGroupRule`) gets called once for each SecurityGroupRule defined (per SecurityGroup).

So exiting here makes sense IMO, if we only want to validate that a SG Rule exists (expecting we wouldn't have duplicates.....which I don't know if the API allows that).

dharaneeshvrd replied:
Yes its trying to match that particular rule passed is exists or not. So exiting as soon as it finds a match.

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+
+// validateVPCSecurityGroup validates the security group and it's rules.
+func (s *PowerVSClusterScope) validateVPCSecurityGroup(securityGroup infrav1beta2.SecurityGroup) (*string, error) {
+	securityGroupDet, err := s.IBMVPCClient.GetSecurityGroupByName(*securityGroup.Name)
```
May we can try to fetch by id as well, if id is already set in status? also we need to make sure name is not nil.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+			return true, fmt.Errorf("error validating existing security group: %v", err)
+		}
+		if securityGroupID != nil {
+			s.Logger.Info("Security group already exists", "name", *securityGroup.Name)
```
Refactored it to get it from `vpcv1.SecurityGroup` where name is a required param.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+				ControllerCreated: ptr.To(false),
+			})
+
+			return false, nil
```
Yes, Done
thanks!

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+		}
+
+		s.SetVPCSecurityGroupID(*securityGroup.Name, infrav1beta2.ResourceReference{
+			ID:                securityGroupID,
```
Please have a look at the latest implementation, added ruleIDs also in status.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+
+	options := &vpcv1.CreateSecurityGroupOptions{
+		VPC: &vpcv1.VPCIdentity{
+			ID: s.GetVPCID(),
```
As per the reconcile order, when it reaches here VPC status would have been populated right?
Do we really need to check? If its not populated, it would requeue on ReconcileVPC only I believe.

dharaneeshvrd commented on pkg/cloud/services/vpc/service.go:
```diff
@@ -352,3 +352,56 @@ func NewService(svcEndpoint string) (Vpc, error) {
+		if start != "" {
+			listSecurityGroupOptions.Start = &start
+		}
+		securityGroupCollection, _, err := s.ListSecurityGroups(&listSecurityGroupOptions)
```
Thanks for the suggestion, implemented using pager.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1841,6 +2166,34 @@ func (s *PowerVSClusterScope) DeleteLoadBalancer() (bool, error) {
+		if _, _, err := s.IBMVPCClient.GetSecurityGroup(&vpcv1.GetSecurityGroupOptions{
+			ID: securityGroup.ID,
+		}); err != nil {
+			if strings.Contains(err.Error(), "not found") {
```
Yes it would, but this method of validating the resource is deletion is followed through out this controller code. 
IMO also it's fine.

dharaneeshvrd commented:
> The design makes sense, handling the complexities of SecurityGroups and SecurityGroupRules definitions.
> 
> In my mind, I was thinking VPC would iterate through creating all the SecurityGroups first, then return to populate the SecurityGroupRules, as some of the Remotes depend on the SecurityGroups (CRN). I will rethink the process some compared with the expectations here, if this looping is acceptable, when to raise errors, etc. due to these additional complexities (inter SecurityGroup/SecurityGroupRule dependencies).

@cjschaef 
I have n't used security groups that much, so could n't think of this use case. With the current implementation if we pass the security groups in order which the dependancy chain is, it would still handle it. But if we really need a freedom of referring the sg's CRN from different order, then we need to redesign and it will bring more complexities than current one.
Please let me know whether the current approach is ok or not!  

Also I have refactored few things from last review, ptal!

dharaneeshvrd commented:
/retest-required

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1129,6 +1156,382 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+// ReconcileVPCSecurityGroup reconciles VPC security group.
+func (s *PowerVSClusterScope) ReconcileVPCSecurityGroup() (bool, error) {
+	for _, securityGroup := range s.IBMPowerVSCluster.Spec.VPCSecurityGroups {
+		securityGroupID, securityGroupRuleIDs := s.GetVPCSecurityGroupID(*securityGroup.Name)
```
Lets carefully handle `*securityGroup.Name` when user sets `securityGroup.ID`

dharaneeshvrd replied:
Sure, taken care, please check!

Karthik-K-N reviewed (commented):
Thanks, Few questions otherwise, Overall LGTM

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1129,6 +1156,382 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+	setRemote := func(remote infrav1beta2.SecurityGroupRuleRemote, remoteOption *vpcv1.SecurityGroupRuleRemotePrototype) error {
+		switch remote.RemoteType {
+		case infrav1beta2.SecurityGroupRuleRemoteTypeCIDR:
+			cidrSubnet, err := s.IBMVPCClient.GetVPCSubnetByName(*remote.CIDRSubnetName)
```
Yes this is validated in kubebuilder rules in spec.

Karthik-K-N reviewed (commented):
LGTM
@mkumatag @Prajyot-Parab @Amulyam24 Please take a look. Thanks

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1129,6 +1169,390 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+					SecurityGroupID: securityGroupID,
+					ID:              rule,
+				}); err != nil {
+					return true, err
```
@dharaneeshvrd, shouldn't we be returning `false` on err and not requeue?

Amulyam24 replied:
The requeue was added for handling the [state of the infra resources](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1693), it can be skipped if security group doesn't need state management.

dharaneeshvrd replied:
Yes Agree, corrected.
Thanks @Amulyam24 !

Amulyam24 replied:
@dharaneesh, since we are returning false every time, maybe we can just return err and remove the requeue check in the controller?

dharaneeshvrd replied:
Corrected!

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1841,6 +2265,34 @@ func (s *PowerVSClusterScope) DeleteLoadBalancer() (bool, error) {
+		if _, _, err := s.IBMVPCClient.GetSecurityGroup(&vpcv1.GetSecurityGroupOptions{
+			ID: securityGroup.ID,
+		}); err != nil {
+			if strings.Contains(err.Error(), "not found") {
```
Done

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1841,6 +2265,34 @@ func (s *PowerVSClusterScope) DeleteLoadBalancer() (bool, error) {
+		}); err != nil {
+			if strings.Contains(err.Error(), "not found") {
+				s.Info("VPC security group successfully deleted", "ID", securityGroup.ID)
+				return false, nil
```
Yes true, I followed the vpc subnet deletion, but I think its bug to return from the loop after processing one item [here](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/3a144b58cc0cc8a14be8babea208c7f2f4aa9e56/cloud/scope/powervs_cluster.go#L1905) ;)

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1129,6 +1169,390 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+			return false, fmt.Errorf("error validating existing security group: %v", err)
+		}
+		if securityGroupDet != nil {
+			s.Logger.Info("Security group already exists", "name", *securityGroupDet.Name)
```
I think its ok here, since its going to this block only one time. Same is there on other components too when user provided spec is already exists in cloud, wdys?

Amulyam24 replied:
By default, we are setting the log level to 3 to keep uniformity as per the new changes in [PR](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1755).

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1166,6 +1206,390 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+			ControllerCreated: ptr.To(true),
+		})
+	}
+
```
Can we log a statement on successful creation of a security group?

dharaneeshvrd replied:
Done

Amulyam24 reviewed (commented):
@dharaneeshvrd, couple of things, with [log refactoring changes](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1755), do you mind following the below in this PR:

- consistent naming for VPC and infra resources in logs such as VPC, VPC subnet, VPC security group etc.
- Error messages starting with `failed to...`

Apart from minor comments, overall LGTM

dharaneeshvrd commented:
@Amulyam24 
I have refactored the logs and changed them to V(3).
Now I am not seeing any logs by default, let me know if you want to change the level here to display some if you planned, or please take care of them in your log refactor PR!
Thanks!

Amulyam24 commented:
> @Amulyam24 I have refactored the logs and changed them to V(3). Now I am not seeing any logs by default, let me know if you want to change the level here to display some if you planned, or please take care of them in your log refactor PR! Thanks!

Sure @dharaneeshvrd, I can take care of them in the logging PR once this is merged. Changes LGTM!
@Prajyot-Parab @mkumatag PTAL.

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1166,6 +1206,393 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+		var protocol string
+		var portMin, portMax int64
+
+		switch direction {
+		case infrav1beta2.SecurityGroupRuleDirectionInbound:
+			protocol = string(expectedRule.Source.Protocol)
+			portMin = int64(expectedRule.Source.PortRange.MinimumPort)
+			portMax = int64(expectedRule.Source.PortRange.MaximumPort)
+			icmpCode := expectedRule.Source.ICMPCode
+			icmpType := expectedRule.Source.ICMPType
+
+			for _, remote := range expectedRule.Source.Remotes {
+				id, match, err := s.validateSecurityGroupRule(originalSecurityGroup, direction, protocol, portMin, portMax, icmpCode, icmpType, remote)
+				if err != nil {
+					return nil, false, fmt.Errorf("failed to validate security group rule: %w", err)
+				}
+				if !match {
+					return nil, false, nil
+				}
+				ruleIDs = append(ruleIDs, id)
+			}
+		case infrav1beta2.SecurityGroupRuleDirectionOutbound:
+			protocol = string(expectedRule.Destination.Protocol)
+			portMin = int64(expectedRule.Destination.PortRange.MinimumPort)
+			portMax = int64(expectedRule.Destination.PortRange.MaximumPort)
+			icmpCode := expectedRule.Destination.ICMPCode
+			icmpType := expectedRule.Destination.ICMPType
+
+			for _, remote := range expectedRule.Destination.Remotes {
+				id, match, err := s.validateSecurityGroupRule(originalSecurityGroup, direction, protocol, portMin, portMax, icmpCode, icmpType, remote)
+				if err != nil {
+					return nil, false, fmt.Errorf("failed to validate security group rule: %v", err)
+				}
+				if !match {
+					return nil, false, nil
+				}
+				ruleIDs = append(ruleIDs, id)
+			}
+		}
```
applies to other functions as well.

dharaneeshvrd replied:
tried to reuse some parts, ptal now!

dharaneeshvrd commented on api/v1beta2/ibmpowervscluster_types.go:
```diff
@@ -103,6 +103,10 @@ type IBMPowerVSClusterSpec struct {
 
+	// securityGroups to attach it to the VPC resource
+	// +optional
+	VPCSecurityGroups []SecurityGroup `json:"vpcSecurityGroups,omitempty"`
```
Since its a list item, named it in plural just like `VPCSubnets` and `LoadBalancers`.

mkumatag replied:
n/m, this `SecurityGroup` struct is created earlier in the types.go file

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1166,6 +1206,393 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
+			securityGroupRuleIDs = s.GetVPCSecurityGroupRuleIDs(*securityGroup.ID)
+		}
+
+		if securityGroupID != nil && securityGroupRuleIDs != nil {
```
Taken care in validate section if its created by controller will retry to create rules.

dharaneeshvrd commented:
Addressed your comments, ptal now @mkumatag

Karthik-K-N reviewed (commented):
Overall LGTM

mkumatag reviewed (approved):
/lgtm
